pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,674 @@
+// obj could be an *v1.VirtualMachine, or a DeletionFinalStateUnknown marker item.
+func (c *OVMController) deleteVirtualMachine(obj interface{}) {
+	vm, ok := obj.(*virtv1.VirtualMachine)
+
+	// When a delete is dropped, the relist will notice a vm in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the vm
+	// changed labels the new ReplicaSet will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
+			return
+		}
+		vm, ok = tombstone.Obj.(*virtv1.VirtualMachine)
+		if !ok {
+			log.Log.Reason(fmt.Errorf("tombstone contained object that is not a vm %#v", obj)).Error("Failed to process delete notification")
+			return
+		}
+	}
+
+	controllerRef := controller.GetControllerOf(vm)
+	if controllerRef == nil {
+		// No controller should care about orphans being deleted.
+		return
+	}
+	ovm := c.resolveControllerRef(vm.Namespace, controllerRef)
+	if ovm == nil {
+		return
+	}
+	ovmKey, err := controller.KeyFunc(ovm)
+	if err != nil {
+		return
+	}
+	c.expectations.DeletionObserved(ovmKey, controller.VirtualMachineKey(vm))
+	c.enqueueOvm(ovm)
+}
+
+func (c *OVMController) addOvm(obj interface{}) {
+	c.enqueueOvm(obj)
+}
+
+func (c *OVMController) deleteOvm(obj interface{}) {
+	c.enqueueOvm(obj)
+}
+
+func (c *OVMController) updateOvm(old, curr interface{}) {
+	c.enqueueOvm(curr)
+}
+
+func (c *OVMController) enqueueOvm(obj interface{}) {
+	logger := log.Log
+	ovm := obj.(*virtv1.OfflineVirtualMachine)
+	key, err := controller.KeyFunc(ovm)
+	if err != nil {
+		logger.Object(ovm).Reason(err).Error("Failed to extract ovmKey from OfflineVirtualMachine.")
+	}
+	c.Queue.Add(key)
+}
+
+func (c *OVMController) hasCondition(ovm *virtv1.OfflineVirtualMachine, cond virtv1.OfflineVirtualMachineConditionType) bool {
+	for _, c := range ovm.Status.Conditions {
+		if c.Type == cond {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *OVMController) removeCondition(ovm *virtv1.OfflineVirtualMachine, cond virtv1.OfflineVirtualMachineConditionType) {
+	var conds []virtv1.OfflineVirtualMachineCondition
+	for _, c := range ovm.Status.Conditions {
+		if c.Type == cond {
+			continue
+		}
+		conds = append(conds, c)
+	}
+	ovm.Status.Conditions = conds
+}
+
+func (c *OVMController) updateStatus(ovm *virtv1.OfflineVirtualMachine, vm *virtv1.VirtualMachine, createErr error) error {
+	// check if we need to update because of appeared or disappeard errors
+	errorsMatch := (createErr != nil) == c.hasCondition(ovm, virtv1.OfflineVirtualMachineFailure)
+
+	// check if we need to update because pause was modified
+	stoppedMatch := ovm.Spec.Running == c.hasCondition(ovm, virtv1.OfflineVirtualMachineRunning)
```

Comment:
The ovm controller currently does not support to be paused. Thant can be removed.